### ISSUE TITLE
EF Core 5 support

### ISSUE THREAD

**Issue by timmi-on-rails** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1088]
.NET 5 is coming and the release is planned for November 2020.
At the same time EF Core 5 will be released.
Today I can already try the preview builds of .NET 5 and EF Core 5 via NuGet.
But I am unable to use the Pomelo.EntityFrameworkCore.MySql adapter, since there are breaking changes in EF Core 5.
I couldn't find a build of the adapter targeting EF Core 5 Preview.

Do you have plans for that and what is the progress?
I saw there is a 5.0-wip branch.

**Comment by lauxjpn** [comment]
An EF Core 5 preview compatible Pomelo release is planned for the second half of June.

**Comment by timmi-on-rails** [comment]
Any news? On what branch are you developing the .net 5.0 upgrade?
Maybe I am able to contribute, to speed things up.
5.0-wip branch hasn't been touched since February.

**Comment by mguinness** [comment]
The current effort is focused on adding spatial and JSON support to 3.x and then afterwards upgrading will be looked at.  Based on previous version upgrades it's a significant amount of work which will take several months based on the limited developer resources.

**Comment by ansarizafar** [comment]
>An EF Core 5 preview compatible Pomelo release is planned for the second half of June.

@lauxjpn  June is almost over. Can we expect An EF Core 5 preview compatible release this month.

**Comment by timmi-on-rails** [comment]
Based on version 3.1.1 I got an upgrade to EF Core 5 preview "working".
I have inserted some records and ran a select statement successfully.

I still have to run the tests and I need help at 2-3 places.
Just wanted to inform you, that it is possible to get a "compiling" version in couple hours.
Probably faster, when knowing EF Core + database provider api.

EF Core 5 does not require net5.0, only netstandard2.1.
Ofcourse this can still change, because it is preview state, still.

**Comment by lauxjpn** [comment]
We will publish an official preview after the JSON support release, of which I have a prototype running at the moment, that still needs a bit of polishing.
So expect an EF Core 5.0 preview compatible release until about 2020-07-15.

**Comment by timmi-on-rails** [comment]
@lauxjpn I am currently working on a project, which is still in an early phase and the first release is planned for end of 2020, so I wanted to update the project dependencies, e.g. efcore to the preview of 5.0, because this will be the up-to-date version, when the project is released.
The reason is to test and find incompatibilities early as possible.
I haven't checked if any new efcore 5 features are useful for the project.
Just planning ahead to stay updated.

**Comment by Simonl9l** [comment]
I believe there is a general hope, across the system developer community, that the teams developing common libraries that are essential components, that in context of the overall developer ecosystem, are this time more broadly lined up with the DotNet Core 5.0 release schedule than they were with the 3.0 release schedule.

Also there are numerous issues on the various .Net Core and more importantly in the case on the EF Core GitHub repo that have a "Fixed in 5.0" conclusion. 

Although 3.1 is supposed to the current LTS release, many of the solutions are rhetorically involving breaking changes or are too complex to be made backward compatible with 3.1, so are not even made candidates for Long Term Support.

As such the develop community are eager to get to use and experiment with the new features, and more specifically, if they have a MySQL dependency, need a Pomelo EF Core 5.0 compatible driver (beta) to even make that use/experiment possible.

We all await with bated breath, and any progress in this regard is greatly appreciated!

**Comment by lauxjpn** [comment]
This finally just got unblocked by #1102 (through #1157 and #1158).
Expect something working by the end of the week.

**Comment by mguinness** [comment]
> The current effort is focused on adding spatial and JSON support to 3.x and then afterwards upgrading will be looked at.

Work on both of these features is now complete, thanks to Laurents for all his hard work!

**Comment by IvanZheng** [comment]
@lauxjpn  When and where can I get a preview release to try on EF Core 5?

**Comment by jparadnikas** [comment]
Hey! Is there a preview release for this? I need to try it, query splitting is essential for me

**Comment by lauxjpn** [comment]
I have already upgraded Pomelo in the way that it compiles with .NET 5. I am currently upgrading the tests, so that they compile as well. Once that is done (probably tomorrow), we will release a first alpha (on our nightly build feed). Once the tests all run, we will release an official preview (probably around the end of the week). We will then introduce a couple of breaking changes and and features.

**Comment by onurh** [comment]
@lauxjpn Is there any change on scheduled time for alpha version?

**Comment by lauxjpn** [comment]
@onurh There is an alpha prerelease in our [nightly build feed](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#nightly-builds). I can also update it with the latest version in a couple of hours. In this latest version, there are still about 100 of the 15,000 tests not working yet, but this is unlikely to be significant. I am actively working on them, to get a version out with _all_ tests running as soon as possible.

**Comment by zaneclaes** [comment]
@lauxjpn I'd like to try this out. I followed the instructions to add the nightly build, but no new versions appear in NuGet. Looking at [the JSON](https://pkgs.dev.azure.com/pomelo-efcore/Pomelo.EntityFrameworkCore.MySql/_packaging/pomelo-efcore-public/nuget/v3/index.json) file which is linked, there is no reference to any version higher than `3.0.0-beta`.

**Comment by zaneclaes** [comment]
Amazing. I now have Pomelo working w/ dotnet5, EFCore5, Identity Server 4, and MySqlConnector 1.0.1.

I had to recreate my databases (couldn't get the IS4 migration to work properly, but no big deal). I have tested all the core identity flows (creating an account, verifying, etc.). But the really nice part is that [many-to-many tables without explicit mappings](https://devblogs.microsoft.com/dotnet/announcing-entity-framework-core-efcore-5-0-rc1/#many-to-many) are now working, which saved me a TON of boilerplate and headache 🎉

**Comment by ezealv** [comment]
> Amazing. I now have Pomelo working w/ dotnet5, EFCore5, Identity Server 4, and MySqlConnector 1.0.1.
> 
> I had to recreate my databases (couldn't get the IS4 migration to work properly, but no big deal). I have tested all the core identity flows (creating an account, verifying, etc.). But the really nice part is that [many-to-many tables without explicit mappings](https://devblogs.microsoft.com/dotnet/announcing-entity-framework-core-efcore-5-0-rc1/#many-to-many) are now working, which saved me a TON of boilerplate and headache 🎉
------------------------------------

How do you make Pomelo work with EF Core 5? Because Pomelo 5 hasn't come out yet.

**Comment by fmauNeko** [comment]
> Will this work with EF Core 5 under DotNet Core 3.1? Since EF Core 5 is supposed to work in this scenario.

Yes, it targets .NET Standard 2.1, as EF Core 5 does. (https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/5.0-wip/src/EFCore.MySql/EFCore.MySql.csproj#L5 https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/5.0-wip/Directory.Build.props#L32)

**Comment by fuzl-llc** [comment]
Hello.  Thanks for all the helpful info on status and for all the hard work!

I added the package source and tried to update from 3.2.3 to the above mentioned alpha and get this:

Unable to find package MySqlConnector. No packages exist with this id in source(s): pomelo-efcore-public

I also tried to use package manager to install it like this:

PM> Install-Package Pomelo.EntityFrameworkCore.MySql -version 5.0.0-alpha.1.20513.5

...and I get the same error.

I also get this warning:

NU1603: Pomelo.EntityFrameworkCore.MySql 5.0.0-alpha.1.20513.5 depends on Microsoft.EntityFrameworkCore.Relational (>= 5.0.0-rc.1.20451.13) but Microsoft.EntityFrameworkCore.Relational 5.0.0-rc.1.20451.13 was not found. An approximate best match of Microsoft.EntityFrameworkCore.Relational 5.0.0-rc.2.20475.6 was resolved.

I just updated to .NET 5 RC 2... so maybe that is causing the warning, but any ideas on the first one?  Any current issues with availability?  Do I need to wait for a future release to test it with .NET 5 RC 2?

**Comment by mguinness** [comment]
@fuzl-llc Did you follow the instructions from [Nightly Builds](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#nightly-builds) as it seems you're missing the entry for nuget.org source where MySqlConnector is located.

**Comment by fuzl-llc** [comment]
That was it, thanks! I had followed the link to 5.0.0-alpha.1.20513.5 given above:

> The latest build [5.0.0-alpha.1.20513.5](https://dev.azure.com/pomelo-efcore/Pomelo.EntityFrameworkCore.MySql/_packaging?_a=package&feed=pomelo-efcore-public&package=Pomelo.EntityFrameworkCore.MySql&version=5.0.0-alpha.1.20513.5&protocolType=NuGet) was just released 2 hours ago.

It has "connect to feed" instructions of its own which do not include the nuget.org source. I'm not sure if Pomelo controls that or not, but if you click that link, then the "Connect to feed" button, then "NuGet.exe," you get the instructions I used...  which apparently are incomplete.

All is working now.  Thanks again!

**Comment by fmauNeko** [comment]
The issue with the "Connect to feed" is that it includes a `<clear />` before the actual feed.
I myself use a nuget.config like that without issue, and without specifying the original nuget feed again.
```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
 <packageSources>
    <add key="Pomelo" value="https://pkgs.dev.azure.com/pomelo-efcore/Pomelo.EntityFrameworkCore.MySql/_packaging/pomelo-efcore-public/nuget/v3/index.json" />
 </packageSources>
</configuration>
```

**Comment by fuzl-llc** [comment]
Interesting.  At one point, I noticed and removed that "clear" (because it cleared out other things I wanted) and it still didn't work for me.  Maybe I had something else misconfigured.  But anyway, the Nightly Builds instructions worked perfectly.

**Comment by mguinness** [comment]
Just an update, Laurents has been slogging away and [5.0.0-alpha.1](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/5.0.0-alpha.1) is now released - thanks for all your hard work, it's appreciated.

**Comment by lauxjpn** [comment]
As @mguinness already said, [5.0.0-alpha.1](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/5.0.0-alpha.1) is out there.

This is basically a release, that is pretty much backward compatible with `3.2.4`, but also forward compatible with EF Core 5.0. The only not yet fully supported 5.0 feature in there are collations, but it still contains our own collation support we introduced for 3.x.
In the next step, we will implement all the planned breaking changes.

At this point in time, all old and new tests by Pomelo and the EF Core team are already green on this alpha release, so we expect it to be quite stable.

The general goal here is to have a stable release out by `2020-11-10`, that contains all planned breaking changes, so everybody can feel confident to use Pomelo with EF Core 5.0 right from the start.

**Comment by baronics** [comment]
Just so I can plan out my next few weeks, how close are you with the 5.0.0 stable release?
As soon as it's posted, I plan to implement it in a MySQL 8.0.20 project that is using 3.2.4. I would rather wait for the breaking changes (I'm guessing with the latest MySqlConnector) before spending time with the alpha.1

**Comment by lauxjpn** [comment]
We will release `5.0.0-alpha.2` tomorrow, which will contain most of the breaking changes that will impact about everyone.
All breaking changes should result in compile-time errors, and they should be obvious to fix when using a modern IDE (if they are not, please let us know; the PRs here usually contain a small sample code section, if something was changed that might impact many people).

After the second alpha release tomorrow, there should only be smaller breaking changes to come (but a couple of nice new features).

However, we do only apply changes with every single test working, so we consider even the alpha versions as stable from a runtime point of view. But a new release can still result in compile time errors (e.g. when we adjust a namespace, add a mandatory parameter, or remove an obsolete feature that has been superseded by another one) that you would need to fix in your code.

So everyone should feel encouraged to start using `5.0.0-alpha.2` once it is out.

**Comment by lauxjpn** [comment]
> I would rather wait for the breaking changes (I'm guessing with the latest MySqlConnector) before spending time with the alpha.1

@baronics The latest MySqlConnector release is already part of `5.0.0-alpha.1` and we don't have had any issues so far. So there are currently no changes planned regarding MySqlConnector in the upcoming releases.

For changes in `5.0.0-alpha.2`, take a look at the [milestones](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/milestone/24?closed=1) (open and closed).

**Comment by jmmccota** [comment]
Net 5 released

**Comment by jmmccota** [comment]
It's possible to use version 3.2.4 with net 5?

**Comment by lauxjpn** [comment]
> It's possible to use version 3.2.4 with net 5?

@jmmccota It is possible to use Pomelo 3.2.4 with .NET 5 and EF Core 3.1.9, because Pomelo is .NET Standard 2.0 compliant (see [Compatibility](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#compatibility)).

But Pomelo 3.2.4 is not compatible with EF Core 5.0.0, so you would need to use our alpha releases or latest [nightly builds](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#nightly-builds) for that.

**Comment by rexebin** [comment]
It is amazing that 5.0.0-alpha.1 just worked, replacing 3.2.4 without any issues. Feels like a stable release. Thank you for the hard work!

**Comment by lauxjpn** [comment]
@rexebin Thanks! That is how it is supposed to be.

> As @mguinness already said, 5.0.0-alpha.1 is out there.
>
> This is basically a release, that is pretty much backward compatible with 3.2.4, but also forward compatible with EF Core 5.0. The only not yet fully supported 5.0 feature in there are collations, but it still contains our own collation support we introduced for 3.x.
>In the next step, we will implement all the planned breaking changes.

So `5.0.0-alpha.1` has intentionally been a release with as little breaking changes but as much backward and forward compatibility as possible. Basically, if you want most of the advantages of EF Core 5, but don't want to make code changes, this is the release to use.

**Comment by lauxjpn** [comment]
[Release](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/5.0.0-alpha.2) `5.0.0-alpha.2` is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/5.0.0-alpha.2).

**Comment by rexebin** [comment]
> [Release](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/5.0.0-alpha.2) `5.0.0-alpha.2` is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/5.0.0-alpha.2).

Thank you! Upgraded, need to fix one line: `UseMySql(connectionString)` to `UseMySql(connectionString, ServerVersion.AutoDetect(connectionString))`, it's up and running.

**Comment by lauxjpn** [comment]
Yes, their are the following options to choose from for the now mandatory `serverVersion` parameter for the `UseMySql()` call:

* Use `new MySqlServerVersion()` to explicitly specify the MySQL server version.
* Use `new MariaDbServerVersion()` to explicitly specify the MariaDB server version.
* Use `MySqlServerVersion.LatestSupportedServerVersion` to explicitly use the latest MySQL server version supported by Pomelo.
* Use `MariaDbServerVersion.LatestSupportedServerVersion` to explicitly use the latest MariaDB server version supported by Pomelo.
* Use `ServerVersion.FromString()` and let the call decided the server version from the supplied string parameter (e.g. `8.0.21-mysql` or `10.5.5-mariadb`).
* Use `ServerVersion.AutoDetect()` and let Pomelo just get the server version directly from the database server.

As in the past, Pomelo will not use features, that are not available in the specified version, but might leverage newer features, if they are supported.

**Comment by baronics** [comment]
Is there a reason you just didn't default to ServerVersion.AutoDetect() and let users overload something else? That way it would not make it a breaking change.

**Comment by lauxjpn** [comment]
> Is there a reason you just didn't default to `ServerVersion.AutoDetect()` and let users overload something else? That way it would not make it a breaking change.

Yes, these are basically the main points behind this decision:

* Using `ServerVersion.AutoDetect()` will open a connection to the database when the `UseMySql()` call is being executed. However, usually with EF Core, a connection is only opened when you actually query the database for the first time (or explicitly call context.Database.OpenConnection()).
The connection opened here will also be a dedicated connection that is unrelated to EF Core, if you call the method with a connection string or a connection object that has not been opened yet. The reason behind that is, that the connection string (by itself or within the unopened connection) might include a database name, that does not exist yet (e.g. because you intend to use `context.Database.EnsureCreated()`; we should rethink, whether this is relevant to unopened connections, but it is relevant to connection strings, which are the go to option for most people).
So the method will remove the database name from the connection string and then open a connection to get the server version. This should work by default, but because of the many different database server implementations, and additional restriction on shared database servers by cloud providers, its also possible that this might just fail if you don't specify your database on some cloud provider instances. And then, depending on where the database is physically located, there might be latency involved for this call, that you want to avoid (e.g. if your database resides in a data center half around the world).
* However, the main reason is that some queries need features like the still quite new `LATERAL` statement (e.g. translated for complex `Include()` and `Where()` calls) or the `ROW_NUMBER() PARTITION BY` clause (could be translated for `SelectMany()` calls). So when you move your app from your state-of-the-art local database instance to your last year remote production database instance, you might suddenly run into unexpected issues.
Since we support MySQL and MariaDB in different versions, and there are also Aurora and other MySQL compatible database server implementations out there, but the supported EF Core functionality might only be a subset for those different implementations, we would like users to make a conscious choice about the database version they want to support.

---

However, if you guy think that there are strong arguments to make auto detection the default, we can also talk about changing the current implementation.

**Comment by codeaid** [comment]
@lauxjpn I'm personally totally with you based on the points you made. I prefer manually configuring things, especially those that can be of many similar yet different types (e.g. as it is with different implementations of MySQL). The application might not even need to connect to the database in certain scenarios for all you know so having to do so only to bootstrap it and read the version seems excessive. If the user does want to do that I don't see a big problem with having to add `ServerVersion.AutoDetect()` to a few lines.

An upgrade to .NET 5.0 is a conscious decision so the fact that code needs to be adjusted in a couple places is probably a very minor inconvenience when migrating to a new major version of the framework.

**Comment by lauxjpn** [comment]
@baronics @codeaid The main issue that I currently have with the `UseMySql()` implementation is its lack of helpful xml docs for the `serverVersion` parameter, which currently looks just like this:

```c#
/// <param name="serverVersion"> The version of the database server. </param>
```

This should be improved, so that its easier to discover how to actually set the `serverVersion` parameter, with `ServerVersion` being `abstract`.

We could also make `MySqlServerVersion` and `MariaDbServerVersion` internal, and expose them (a bit similar to how we did it in 3.x) using `ServerVersion` with a `version` and a `type` parameter. This would mean that all common options would be reachable from the `ServerVersion` class itself.

**Comment by rexebin** [comment]
@lauxjpn Thanks for the information. I totally agree the information given here should go to the documents. I am ok with AutoDetect because I am calling `Migrate()` every time the app starts, so I do access database every time the app starts. 
But after reading your posts, I think I should provide version manually instead of using AutoDetect, one extra trip to database is still one extra. 

I am using Percona which is MySql 8.0 as far as I understand. Is providing major version without the minor versions sufficient?

**Comment by lauxjpn** [comment]
> Is providing major version without the minor versions sufficient?

@rexebin No, the minimum you currently have to provide is the major and minor version. Take a look at the [ServerVersionTest.TestValidVersion](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/83c699f5b747253dc1b6fa9c470f469467d77686/test/EFCore.MySql.Tests/Migrations/ServerVersionTest.cs#L11-L24) test case. It shows working version strings. We will update the opposite test `ServerVersionTest.TestInvalidVersion`, to show that a string of simply `8` or `8-mysql` throws.

I'm currently haven't Percona installed, but would be interested in the result of running a `select @@version;` statement against it, if you got time.

**Comment by rexebin** [comment]
@lauxjpn I ran the query against the Percona docker, it returns 8.0.20-11

**Comment by khteh** [comment]
Any ETA for this?

**Comment by adamhathcock** [comment]
Checkout the milestones.

The issues outstanding don't seem like huge issues for most people.  I do wonder how the maintainers feel of this is ready enough for general use though.

**Comment by lauxjpn** [comment]
@khteh The alpha versions are generally stable. We just call them alpha because we are still introducing breaking changes before the GA release. However, those are mainly breaking changes that will lead to compile-time errors, that should be pretty easy to fix. So feel free to use the alpha releases or the nightly builds.
As for an official GA release date, somewhere before the end of the second December week seems realistic to me.

I am currently revamping the collation and char set support. After that, the first beta shouldn't be far.

**Comment by MarkCiliaVincenti** [comment]
I'm using alpha2 on a project since its release and I haven't run into any problems whatsoever.

**Comment by patil-vaibhav** [comment]
Will there be any problem if character set is different than `utf8mb4` . I'm using 10.1.44- MariaDB Server for production with `utf8_general_ci` .

**Comment by lauxjpn** [comment]
> Will there be any problem if character set is different than `utf8mb4` . I'm using 10.1.44- MariaDB Server for production with `utf8_general_ci` .

5.0.0 should generally handle charsets as before. However, the `UseMySql(..., option => option.CharSet())` method is being made obsolete and will be replaced by a `HasCharSet()` method on the model (like `modelBuilder.HasCharSet()`), the table and the column (it is already available at the column level since 3.x in Pomelo).

The default will be `utf8mb4`, so if you use another character set, you should explicitly set it with one of the `HasCharSet()` options once they are available.

**Comment by barnuri** [comment]
@lauxjpn 
hi, when you think ef core 5 scaffold flags are will be available in Pomelo ? 
I try to use flags --no-onconfiguring, --namespace , --context-namespace and its didnt works.

Also there is issues if your project enabled Nullabled support than the scaffold dont detect the diffrence between string and string?
in the offical package of sqlserver they add for every file of the scaffold the line that disable nullable in all file

```
#nullable disable
public class ExEntity 
{
...
...
...
}
```

**Comment by lauxjpn** [comment]
@barnuri Be so kind and just open an issue about it. I will then fix it shortly.

**Comment by barnuri** [comment]
@lauxjpn 
#1277 
#1276

**Comment by Ghostbxz** [comment]
@barnuri I have the same problem I need a solution for it too

**Comment by mguinness** [comment]
@Ghostbxz Please comment on the respective issue that is affecting you.

**Comment by pkellner** [comment]
How can I get an EF context object that I can use with a .net 5 console app (using the alpha 2).  That is, I get how to create this in a web app

```
services.AddDbContextPool<YourDbContext>(
                dbContextOptions => dbContextOptions
                    .UseMySql(
                        // Replace with your connection string.
                        "server=localhost;user=root;password=1234;database=ef",
                        // Replace with your server version and type.
                        // For common usages, see pull request #1233.
                        new MySqlServerVersion(new Version(8, 0, 21)), // use MariaDbServerVersion for MariaDB
                        mySqlOptions => mySqlOptions
                            .CharSetBehavior(CharSetBehavior.NeverAppend))
                    // Everything from this point on is optional but helps with debugging.
                    .EnableSensitiveDataLogging()
                    .EnableDetailedErrors();
            ));
```

Then in my controller inject the service I assume and get a context, something like this I'm thinking

```
   public class DiscountCodesController : ControllerBase
    {
        private readonly IDbContextFactory<ApplicationDbContext> _contextFactory;
        public DiscountCodesController(IDbContextFactory<ApplicationDbContext> contextFactory)
        {
            _contextFactory = contextFactory;
        }
        
        [HttpGet("/rest/discountcodes")] 
        public  object Get()
```

But I can't figure out (even looking at the test examples), How, in a console app I can directly get a context using the factory pattern.

**Comment by barnuri** [comment]
@pkellner 
you need to inject the DbContext in your controller
your injection in services, services.AddDbContextPool... inject the dbcontext to the DI, and use it as a Pool
you also can add the injection what the lifecycle for the injection
```
public class DiscountCodesController : ControllerBase
    {
        private readonly ApplicationDbContext  _context;
        public DiscountCodesController(ApplicationDbContext context)
        {
            _context = context;
        }
        
        [HttpGet("/rest/discountcodes")] 
        public  object Get() {
            var z = context.DiscountCodes
                .Select(dc => new
                {
```

**Comment by Ghostbxz** [comment]
> @Ghostbxz Please comment on the respective issue that is affecting you.

done

**Comment by lauxjpn** [comment]
> How can I get an EF context object that I can use with a .net 5 console app (using the alpha 2). [...]

@pkellner Here is a simple .NET 5.0 console project, using EF Core 5.0.1 and Pomelo 5.0.0-alpha.2:

<details>
<summary>Project file</summary>

```xml
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net5.0</TargetFramework>
        <Nullable>disable</Nullable>
    </PropertyGroup>

    <ItemGroup>
        <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="5.0.0" />
        <PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="5.0.0-alpha.2" />
    </ItemGroup>

</Project>
```

</details>

<details>
<summary>Program.cs</summary>

```c#
using System.Diagnostics;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Pomelo.EntityFrameworkCore.MySql.Infrastructure;

namespace IssueConsoleTemplate
{
    public class IceCream
    {
        public int IceCreamId { get; set; }
        public string Name { get; set; }
    }

    public class Context : DbContext
    {
        public DbSet<IceCream> IceCreams { get; set; }

        // Using this contstructor instead of OnConfiguring is also an option:
        //
        // public Context(DbContextOptions<Context> options)
        //     : base(options)
        // {
        // }
        
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            var connectionString = "server=127.0.0.1;port=3306;user=root;password=;database=Issue1088";
            
            optionsBuilder.UseMySql(
                    connectionString,
                    ServerVersion.AutoDetect(connectionString),
                    options => options.CharSetBehavior(CharSetBehavior.NeverAppend))
                .UseLoggerFactory(
                    LoggerFactory.Create(
                        configure => configure
                            .AddConsole()
                            .AddFilter(level => level >= LogLevel.Information)))
                .EnableSensitiveDataLogging()
                .EnableDetailedErrors();
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<IceCream>(
                entity =>
                {
                    entity.Property(e => e.Name)
                        .HasMaxLength(255)
                        .IsRequired();

                    entity.HasData(
                        new IceCream {IceCreamId = 1, Name = "Vanilla"},
                        new IceCream {IceCreamId = 2, Name = "Chocolate"});
                });
        }
    }

    internal static class Program
    {
        private static void Main(string[] args)
        {
            using var context = new Context();

            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();

            var iceCreams = context.IceCreams
                .OrderBy(i => i.IceCreamId)
                .ToList();

            Trace.Assert(iceCreams.Count == 2);
            Trace.Assert(iceCreams[0].Name == "Vanilla");
            Trace.Assert(iceCreams[1].Name == "Chocolate");
        }
    }
}
```

</details>

**Comment by Noplace** [comment]
any target to update to .NET 5.0? I want to update my EF packages but not sure if I can/should while pomelo is still on Core 3.1

**Comment by lauxjpn** [comment]
> any target to update to .NET 5.0? I want to update my EF packages but not sure if I can/should while pomelo is still on Core 3.1

@Noplace You can update you projects to .NET 5 at any time you want.

EF Core 3.1.x and Pomelo 3.2.x are both .NET Standard 2.0 compliant, so they full support .NET (Core) 2.0+ (so including .NET 5) and  .NET Framework 4.6.1+.

EF Core 5.0.x and Pomelo 5.0.0-alpha.2 are both .NET Standard 2.1 compliant, so they full support .NET (Core) 3.0+ (so including .NET 5).

For more information, see our [compatibility matrix](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/#compatibility).

**Comment by Noplace** [comment]
> > any target to update to .NET 5.0? I want to update my EF packages but not sure if I can/should while pomelo is still on Core 3.1
> 
> @Noplace You can update you projects to .NET 5 at any time you want.
> 
> EF Core 3.1.x and Pomelo 3.2.x are both .NET Standard 2.0 compliant, so they full support .NET (Core) 2.0+ (including .NET 5) and .NET Framework 4.6.1+.
> 
> EF Core 5.0.x and Pomelo 5.0.0-alpha.2 are both .NET Standard 2.1 compliant, so they full support .NET (Core) 3.0+ (including .NET 5).
> 
> For more information, see our [compatibility matrix](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/#compatibility).
> 
> In case you meant "EF Core 5.0.1" instead of ".NET 5.0", you will need to use the latest _prerelease_ of Pomelo (currently `5.0.0-alpha.2`).

thank you for your reply, the issue is I also have other packages which are causing conflicts, mainly the Hangfire libraries which for MySql depend on MySqlConnector package being version > 1.0 while Pomelo requires MySqlConnector (>= 0.69.9 && < 1.0.0)

was hoping the dependency requirement would get updated instead of working around it

**Comment by lauxjpn** [comment]
> was hoping the dependency requirement would get updated instead of working around it

Pomelo `5.0.0-alpha.2` uses MySqlConnector `1.1.0` (see [compatibility matrix](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/#compatibility)). So there should be no need to explicitly use two MySqlConnector versions anymore in your case.

**Comment by Noplace** [comment]
> > was hoping the dependency requirement would get updated instead of working around it
> 
> @Noplace Pomelo `5.0.0-alpha.2` uses MySqlConnector `1.1.0` (see [compatibility matrix](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/#compatibility)). So there should be no need to explicitly use two MySqlConnector versions anymore in your case.

Thank you for pointing that out, everything is sorted now!

**Comment by passuied** [comment]
hi, just a follow up when we should expect the GA release. The original estimate was end of December...
Any blocker for making 5.0.0 stable?
Thanks in advance

**Comment by mguinness** [comment]
There are 25 outstanding issues in the [5.0.0 Milestone](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/milestone/16) so it's not imminent.

**Comment by lauxjpn** [comment]
There has been a [bug discovered](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1302), where the collation you setup for you model (either using the Pomelo specific `.HasCollation()` or the new .NET 5 `.UseCollation()` method) is not being honored in `5.0.0-alpha.2`.

We released a quick [fix](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/pull/1307), that is already available as part of the latest [nightly build](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#nightly-builds) (as a _prerelease_).

**Comment by MarkCiliaVincenti** [comment]
> There has been a [bug discovered](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1302), where the collation you setup for you model (either using the Pomelo specific `.HasCollation()` or the new .NET 5 `.UseCollation()` method) is not being honored in `5.0.0-alpha.2`.
> 
> We released a quick [fix](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/pull/1307), that is already available as part of the latest [nightly build](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#nightly-builds) (as a _prerelease_).

Not sure whether or not this is intentional, but just in case it's not kindly note that alpha3 has not been posted on nuget.org yet.

**Comment by codeaid** [comment]
@MarkCiliaVincenti I'm pretty sure it's because there is still one outstanding issue in the [5.0.0-alpha.3 milestone](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/milestone/26).

**Comment by lauxjpn** [comment]
> Not sure whether or not this is intentional, but just in case it's not kindly note that alpha3 has not been posted on nuget.org yet.

@MarkCiliaVincenti `5.0.0-alpha.3` has not been released yet. As stated, you can use the latest nightly build though:

> We released a quick [fix](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/pull/1307), that is already available as part of the latest [nightly build](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#nightly-builds) (as a _prerelease_).

**Comment by Simonl9l** [comment]
Hi - do we have a status page or something we easily can find, that show what works in DotNet 5 and where the gaps are?

**Comment by lauxjpn** [comment]
@Simonl9l If you use the latest [nightly build](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#nightly-builds), there should be no gaps in regards to .NET 5 and not gaps in regards to EF Core 5.
So if you find any, please just report them and they will be fixed immediately.

About features planned for 5.0 (though not all of them will be in the first official release), just [filter issues by milestone](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues?q=is%3Aopen+is%3Aissue+milestone%3A5.0.0).

**Comment by renemadsen** [comment]
@lauxjpn could it be considered to move some of the issues from 5.0.0 into a 5.1.0? to "just" make a net 5.0 compatible version, and move new features into the 5.1.0 milestone?

At least just for making it more clear to outsiders what is blocking a release which is supporting net 5.0.

Just my thoughts, since the list include issues which is dated back to 2016.

**Comment by lauxjpn** [comment]
@renemadsen The only real breaking change coming is reorganization of the charset and collation support, which I will finish as soon as I have a bit of time on my hands again. For most people (those who have been implicitly or explicitly been using UTF-8), they will either need to change one line of code, or none at all.

There are one or two smaller breaking changes (e.g. making the `StringComparison` support opt-in), but they will be insignificant for most.

After that, there will be a beta and a closely followed release. Most features currently on the 5.0.0 milestone will be released before 6.0.0, but not with the 5.0.0 release.

As said before, feel encouraged to use the latest [nightly build](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#nightly-builds), as it is stable.

**Comment by agmorpheus** [comment]
I am using the alpha2. Maybe this is a general problem, but I have the problem, that the API Server takes long to connect to the MYSQL Server (installed on the same server). The Api returns after 15sec a result. In the log I find this:
`An error occurred using the connection to database '' on server '127.0.0.1'.`

Can someone help me with this?

**Comment by lauxjpn** [comment]
@agmorpheus I created #1328 for your issue.

**Comment by lauxjpn** [comment]
I will have more time on my hands again over the next days and weeks, so we will finally get the GA release out the door.

**Comment by funkel1989** [comment]
hey @lauxjpn sorry to be a pest but just curious when your next alpha release is going to be out? Unfortunately GitHub actions, my CICD choice, can only support a single nuget entry (unless anyone can provide a solution to this) so if i register to pull down your nightly builders everything else fails to restore and vise versa. Its become slightly problematic. Really appreciate your work and sorry again for pestering :)

**Comment by Simonl9l** [comment]
The alpha-2 is also now behind on SDK dependencies.

 Is it at all possible to push out another interim alpha/beta/RC before GA release on nugget if this is still not imminent, that would be appreciated!

**Comment by Zwergenpunk** [comment]
Honest question: Is that project goind to die or will you maintain it for dotnet 6?

**Comment by adamhathcock** [comment]
I know people are waiting for this, I am too, but we don't need to ask silly questions.

There's only one question to ask: "how can I help?"

**Comment by renemadsen** [comment]
> I know people are waiting for this, I am too, but we don't need to ask silly questions.
> 
> There's only one question to ask: "how can I help?"

I think the reason for the question is, that net 6.0 is due soon and an update for net 5.0 has not yet seen the light. Yes I know there is the preview. But it's not the same.

I would suggest, that in future releases, they follow net release cycle.
Anything new will come second and be cut out from the new release till after update to accomedate net changes if that makes sense?

Better to release a lot of small changes than wait for a big bang and people start to create new alternatives caused by lack of releasing for the latest net version.

The current list of what is blocking is hugh. I suggest to move all none critical blocking issues to a new milestone, so it's clear, what is missing. I for one just saw the list and was wondering how is that ever going to be done, with issues dating back to 2016?

I gave up finding an issue I could help with, since the list was so long. So I understand the question completely.

**Comment by kampilan** [comment]
Silly question? How is this a silly question? This project has all the earmarks of death. This project needs to taken over or forked by someone with a plan to monetize it. It is to important to let die and to complicated to be developed by a few brave volunteers. In the last 4 months there has been commits by virtually 1 person. A single brave soul probably not making a dime off his/her efforts. This project has a long history of trouble staying updated. There have been calls in the past for new people to take over. New people took over with the same result. Clearly this is a very complicated project requiring large amounts of time to develop and maintain. A different model is needed. I would gladly pay $50 a month to continue to use it and I suspect there are probably plenty more willing to do the same. I am not criticizing the people currently working on the project, I am grateful for their efforts. It just isn't working. Amazon has a big stake in in the MySql/Maria/Aurora eco-system via RDS. Does anyone have any ideas on approaching them for sponsorship? A know Microsoft has been approached in the past and I think they may have pitched in, but they don't perceive themselves as having a stake in the project.

**Comment by lauxjpn** [comment]
> Honest question: Is that project goind to die or will you maintain it for dotnet 6?

It is not going to die. It is healthy and active. We will start targeting EF Core 6, as soon as our 5 GA release is out.

> There's only one question to ask: "how can I help?"

Fixing issues is always appreciated. Even better would be help on the documentation part, because we are lacking there.

---

I have been working on the beta release for about a week and it is pretty much done. The current state can bee seen on my fork's [feature/charset_and_collation](https://github.com/lauxjpn/Pomelo.EntityFrameworkCore.MySql/tree/feature/charset_and_collation) feature branch.
I expect it's push today. The `5.0.0-beta.1` will be out about 24 hours after that.

If there are no serious issues discovered within about a week, the GA release will be published.

**Comment by renemadsen** [comment]
> Fixing issues is always appreciated. Even better would be help on the documentation part, because we are lacking there.

Yes, but could we triage the list of issues, and move to milestone 6.0 7.0 etc? I'll gladly help with that if wanted?

**Comment by lauxjpn** [comment]
> I would suggest, that in future releases, they follow net release cycle.
>Anything new will come second and be cut out from the new release till after update to accomedate net changes if that makes sense?

It's generally a good plan. However, it will either force us to break with SemVer2 compliance (if we release breaking changes within a major release), with independent version numbers from EF Core (if we keep SemVer2 compliance and increase the major version number for breaking changes) or with postponing (possibly needed) features to the next EF Core release.

Which is probably fine, just putting it out there.

> Yes, but could we triage the list of issues, and move to milestone 6.0 7.0 etc? I'll gladly help with that if wanted?

A couple of the features in the list have already been fixed or will be fixed once the latest feature branch is pushed. I'll update them when I release the beta tomorrow.

I'll increase the milestone for features with breaking changes to the next major one, once the GA release is out.

If you want to tackle a feature, just say so on the issue page, so we can coordinate.

> Silly question? How is this a silly question? This project has all the earmarks of death. This project needs to taken over or forked by someone with a plan to monetize it. It is to important to let die and to complicated to be developed by a few brave volunteers. In the last 4 months there has been commits by virtually 1 person. A single brave soul probably not making a dime off his/her efforts. This project has a long history of trouble staying updated.

Well, technically I am basically the only significantly changing the codebase for about 1,5 years. Generally, that is fine with me, but I had a crazily busy year due to the pandemic (most of my clients are local governments, which suddenly had the need to move their decades-old paper-based processes to IT-based ones).
Luckily, most of this work is done now.

As for the current `alpha` state, we only named it `alpha`, because of the (two?) breaking changes, that were planned for it. The quality is the same level as our major releases. As stated in many posts above, anybody who wants to use it in production, should feel free to do so (though its probably a good idea to subscribe to this thread, as I will post about the breaking changes here as well).

---

Of course it is always preferable to have more people committing time to the project, but as long we are not being hit with a different significant global issue, I think its fine.

If you want to discuss methods, release planning, EF Core release coordination etc., feel free to open an issue about it, so we track it properly.

**Comment by lauxjpn** [comment]
Regarding the upcoming beta release, it will be the one with the breaking changes, where we made the `DbContext` option `CharSet()` obsolete and moved everything charset and collation related to the model definition (Fluent API).

The defaults on the model level will result in basically the same SQL generation of the `DbContext` option we used previously. But if you were explicitly using the `CharSet` option with a different setting than `utf8mb4` before, then you should define the charset now on the model level (and not explicitly disable its delegation).

You will be able to specify charsets and collations on the model/database, the entity/table and the property/model level and we support it in when scaffolding or creating migrations as well.

**Comment by mguinness** [comment]
> > There's only one question to ask: "how can I help?"
> 
> Fixing issues is always appreciated. Even better would be help on the documentation part, because we are lacking there.

Another way to help is to [watch](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/watchers) the repo and help with answering [issues](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues) and also on [Stack Overflow](https://stackoverflow.com/questions/tagged/pomelo).  That is a relatively easy way to get involved and eases the burden on the main developer, meaning they can spend more time coding and less time on support.

**Comment by renemadsen** [comment]
> It's generally a good plan. However, it will either force us to break with SemVer2 compliance (if we release breaking changes within a major release), use independent version numbers from EF Core (if we keep SemVer2 compliance and increase the major version number for breaking changes) or to postpone (possibly needed) features to the next EF Core release.

It makes sense, but if we assume, that breaking changes could be identified in the development phase of EF Core next version, then next version would include the needed breaking changes. 

Other self introduced breaking changes would be possible to postpone to next version release and hence be ready at the time next version of EF core is about to be released.

> I'll increase the milestone for features with breaking changes to the next major one, once the GA release is out.

Sounds good, could that include 6.0 7.0? So we who would like to help could know, which ones is thought of being really needed for next version, so the release for net core 6.0 could be ready faster? Like almost at the time when net 6.0 is ready.

What I'm asking is simply some filtering of issues, so the project could move on and not get stopped because new issues of interest keeps being added to the current milestone if that makes sense?

**Comment by lauxjpn** [comment]
> Sounds good, could that include 6.0 7.0? So we who would like to help could know, which ones is thought of being really needed for next version, so the release for net core 6.0 could be ready faster? Like almost at the time when net 6.0 is ready.

I would generally implement new features for the new major release and fixes of issues as a servicing release on the latest stable branch. So it makes sense to update the outstanding features then to the `6.0.0` milestone.

Usually we only plan as far as the next major release. So something is either planned for the next service release, the next minor release, the next major release or in the backlog (so there would be no planning for 7.0.0 at the moment, but only 6.0.0).

I agree that should define a feature cut-off at least a month before the next major EF Core version release, so we can deliver on time. We would then increase the milestone version on the features that didn't made the cut. (As for the current issues, I will go over it once the GA release is out.)

While we had plans for this last year as well, we plan to target the EF Core previews once 5.0.0 GA is out, so we can release on time with the EF Core 6 when it goes GA.

**Comment by Zwergenpunk** [comment]
> > Honest question: Is that project goind to die or will you maintain it for dotnet 6?
> 
> It is not going to die. It is healthy and active. We will start targeting EF Core 6, as soon as our 5 GA release is out.
> 

That's good news.
Just to clarify, I did not want to criticize you or the work you're doing. I just had the feeling that there is a possibility that your priorities had shifted too much (which would be fine, just good to know)

I did not excpect this to blow up, but I'm happy that alot of  questions could be answered.

Thanks for your work.

**Comment by lauxjpn** [comment]
> Just to clarify, I did not want to criticize you or the work you're doing. I just had the feeling that there is a possibility that your priorities had shifted too much (which would be fine, just good to know)

We are basically half a year later than we wanted to be with our GA release, so it is a legitimate question. But as long as there are pull requests being merged, the project is probably healthy.

Personally, I don't think it is going anywhere soon. It has a much stronger community then Oracle's implementation, is faster, more versatile, contains many major features that are not part of Oracle's implementation and has a quick reaction time to issues posted here or on StackOverflow.

**Comment by mc0re** [comment]
> Regarding the upcoming beta release, it will be the one with the breaking changes, where we made the `DbContext` option `CharSet()` obsolete...
>
> ...the second (much smaller) breaking change, that is about the case in-/sensitive comparison translations

There is one more breaking change that I hit - removal of support for provider-specific functions from _In-memory_ database provider. Now when running against `InMemory`, things like `Hex()` throw.

**Comment by lauxjpn** [comment]
> There is one more breaking change that I hit - removal of support for provider-specific functions from _In-memory_ database provider. Now when running against `InMemory`, things like `Hex()` throw. See #1355 for details.

For a list of all general EF Core 5.0 breaking changes, see [Breaking changes in EF Core 5.0](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-5.0/breaking-changes).

**Comment by lauxjpn** [comment]
> I expect it's push today. The `5.0.0-beta.1` will be out about 24 hours after that.

PR #1385 has been pushed now.

**Comment by amin-sage** [comment]
Hi, I still can't see Beta version in releases, is it out already?

**Comment by lauxjpn** [comment]
> Hi, I still can't see Beta version in releases, is it out already?

As said:

> The `5.0.0-beta.1` will be out about 24 hours after [the push].

So, expect tomorrow. I did also update the

**Comment by lukas-kamleitner** [comment]
Will support for JSON operations on MySQL will be part of the 5.0.0 release?
In the current version 3.2.5 `EF.Functions.JsonExtract` seems not to be implemented

**Comment by mguinness** [comment]
Have you tried using it in 5.0?  Look at [JsonPocoQueryTestBase](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/test/EFCore.MySql.FunctionalTests/Query/JsonPocoQueryTestBase.cs) for examples.  Also see https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/14#issuecomment-725257107.

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/d48240aa92f918c5f27f377230b02fd862c3dbaf/src/EFCore.MySql/Query/ExpressionTranslators/Internal/MySqlJsonDbFunctionsTranslator.cs#L84-L92

**Comment by lukas-kamleitner** [comment]
Already got it working, thanks. Sorry for not opening a new ticket for the question.

**Comment by lauxjpn** [comment]
[Release](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/5.0.0-beta.1) `5.0.0-beta.1` is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/5.0.0-beta.1).

As [stated above](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1088#issuecomment-822557753), it contains breaking changes about how Pomelo handles the default character set (also see #1385).

**Comment by lauxjpn** [comment]
We will release `5.0.0-beta.2` on Sunday, that we expect to contain the last changes before the GA release on Wednesday.

Notably, there where two issues related to scaffolding (#1393 and #1394), that will be fixed in it.
(As always, take a look at the [milestone](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/milestone/28) for everything related.)

**Comment by lauxjpn** [comment]
[Release](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/5.0.0-beta.2) `5.0.0-beta.2` is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/5.0.0-beta.2).

If you need the _exact_ charset behavior as in previous versions for any reason, use the following Fluent API code:

```c#
// The following applies the "utf8mb4" charset recursively only to columns, without changing the default charset of
// the database or its tables.
modelBuilder.HasCharSet("utf8mb4", DelegationModes.ApplyToColumns);

// The following disables the newly introduced default collation 'ascii_general_ci' for character based Guid columns.
modelBuilder.UseGuidCollation(string.Empty);
```

Otherwise, just use the default (`utf8mb4` all the way through, except for character based `Guid` properties, which now use the `ascii_general_ci` collation and thus the `ascii` charset) by not calling anything explicitly, or setup precisely what you need.

Special thanks to @mc0re for his feedback, which significantly helped to improve this release and iron out the remaining issues.

**Comment by mstaal** [comment]
I am currently using ASP.NET Core Identity based on usage of your software. I have created my MySQL tables using statements like below. Is it correct that I should alter the type of `Id`, and is there an obvious way of changing the CREATE statement? Can you refer to a reason why the changes were made?
```
CREATE TABLE IF NOT EXISTS AspNetRoles (
  `Id` varchar(128) NOT NULL,
  `Name` varchar(256) NOT NULL,
  `NormalizedName` varchar(256) NOT NULL,
  `ConcurrencyStamp` varchar(256) DEFAULT NULL,
  PRIMARY KEY (`Id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
```

**Comment by khteh** [comment]
I don't think you need to change anything there. `ascii_general_ci` means that it will use exactly 128 bytes to store the GUID instead of variable 1 to 4 bytes if using unicode, for instance.

**Comment by mstaal** [comment]
Thanks, @khteh! Lovely to know. I look much forward to support for (EF Core) v5!

**Comment by khteh** [comment]
But @mstaal why `varchar(128)` for GUID? I would use either `varchar(36)` or `BINARY(16)` for GUID.

**Comment by mstaal** [comment]
@khteh : To be honest I must admit that I once found inspiration in [this gist](https://gist.github.com/jambelnet/1ea70236c933d644b36895b296fb44e5), which is not a very mature explanation. It is a good point that I could perhaps benefit from changing that. I always welcome feedback.

**Comment by lauxjpn** [comment]
> I don't think you need to change anything there. `ascii_general_ci` means that it will use exactly 128 bytes to store the GUID instead of variable 1 to 4 bytes if using UTF8, for instance.

You mean 128 _bits_ (16 bytes), but storage space is indeed one of the tree main reasons for the change.

The second one is performance: the `ascii` charset has the best performance (due to less storage space and minimal collation logic).

The third one is making sure, that users don't accidentally use a case sensitive collation. This might not be strictly necessary, if all GUIDs are always newly created, as they don't collide. But if for some reason someone tries to insert an existing entity, but with a differently cased GUID into to table, we want to make sure that this leads to an error by default. If this cannot happen to a user's app, performance could be improved slightly by manually using a CS collation instead (though if it is max. performance you are after, then using `binary(16)` is the way to go).

> [...] I look much forward to support for (EF Core) v5!

Pomelo `5.0.0` GA is going to be released later today. We are just waiting a bit to make sure, that nobody encounters any serious last minute issues.

**Comment by lauxjpn** [comment]
[Release](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/5.0.0) `5.0.0` is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/5.0.0).

Thanks to everybody for their patience and special thanks to anybody who contributed!

**Comment by passuied** [comment]
thanks for making it happen!

**Comment by mstaal** [comment]
Thanks for the update! I just deployed my .NET 5 version of my small user management system, and everything seems to work just fine!

**Comment by LeaFrock** [comment]
So ... EF Core 6 Support is on the way, haha  :beers:

**Comment by lauxjpn** [comment]
> So ... EF Core 6 Support is on the way, haha 🍻

Absolutely, I started working on it immediately after the 5.0 GA release, to target 6.0.0-preview4. See #1413 for upcoming information.
